### ISSUE TITLE
rancher-cis-benchmark k8s 1.27 chart support (Infracloud)

### ISSUE THREAD

**Issue by gunamata** [rancher/rancher#41787]
This issue is to implement the chart changes & perform the testing required to upgrade to k8s 1.27. Please perform below steps. 

- [x] Check whether the chart has any references to the below deprecated features of k8s 1.27. You can find information about deprecated APIs here on [Kubernetes Deprecation guide](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-27).

**storage.k8s.io/v1beta1** API version of **CSIStorageCapacity**
 

- [x] Update kube-version annotation in Chart.yaml. catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.28.0-0'

Note: If this is an upstream chart and if the upstream chart doesn't support k8s 1.27 then do not update the kube-version annotation or update the k8s version accordingly.

- [ ] Perform below tests

- Fresh install/uninstall/upgrade/downgrade of the charts on K8s 1.27 provisioned on Rancher. Test the chart functionality.
(Please refer to the https://github.com/rancher/rancher/issues/41395 to know what testing was done for k8s 1.26 upgrade for your chart)
- Provision rancher on k8s 1.26 or lesser
- Install chart on this cluster
- Upgrade this cluster in rancher to k8s 1.27
- Make sure the existing chart and the app is functional after cluster upgrade

**Comment by JonCrowther** [comment]
Need to test the scan of the following clusters:

- [ ] RKE2
- [ ] RKE2 Hardened
- [ ] k3s 
- [ ] k3s Hardened
- [ ] Custom
- [ ] Imported
- [ ] RKE1
- [ ] RKE1 Hardened

**Comment by JonCrowther** [comment]
@Priyashetty17 I will be taking this over FYI. If you've already started could you please post any work you've already done? Thanks

**Comment by rayandas** [comment]
So this issue is about adding support for CIS 1.27 and 1.28 or adding Kubernetes version 1.27 and 1.28 in the `Chart.yaml` kube-version annotation in https://github.com/rancher/charts/blob/623bea830a780e913b4394864d167b98441be4b8/charts/rancher-cis-benchmark/4.2.0-rc2/Chart.yaml#L5 ?

**Comment by gunamata** [comment]
@rayandas , This ticket is about updating the chart to support k8s 1.27. Please note we went a bit back and forth on whether this ticket covers support for just k8s 1.27 or both k8s 1.27 and 1.28. It's finally decided that we will target only k8s 1.27 for the [2023-Q4-v2.8x](https://github.com/rancher/rancher/milestone/315) release.

**Comment by mitulshah-suse** [comment]
@JonCrowther @andypitcher Will you guys be handling this ticket? Or you want us (infracloud team) to handle this?

**Comment by samjustus** [comment]
@mitulshah-suse we have a ton on our plate for Q4 - if you guys could pick it up that would be awesome

**Comment by mitulshah-suse** [comment]
Sure, will pick it up.

**Comment by samjustus** [comment]
@rayandas @mitulshah-suse hey guys - following up on this issue - can you try and get this done ASAP? it was due last week. sorry for not following up earlier

**Comment by mitulshah-suse** [comment]
@samjustus PR is already up. https://github.com/rancher/charts/pull/2839 will merge it soon. We cannot test it until 1.27 support is added for rancher though.

**Comment by rayandas** [comment]
@samjustus Merged CIS-Benchmark PR.

**Comment by Priyashetty17** [comment]
**QA TEST PLAN**

<table>
    <tbody>
        <tr>
            <th>#</th>
            <th>Scenario</th>
            <th>Steps</th>
            <th>Expected</th>
        </tr>
        <tr>
            <td>1</td>
            <td>Fresh install Rancher with hardened downstream RKE1 cluster - k8s v1.27 and CIS profile v1.23</td>
            <td>
                <ol>
                    <li> Create a rancher server v2.8-head. </li> 
                    <li> Create a custom hardened downstream RKE1 cluster with k8s v1.27 and CIS profile v1.23. </li>
                    <li> Install "rancher-cis-benchmark" chart on the cluster. </li>
                    <li> Verify kube-version annotation in Chart's yaml: catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.28.0-0' </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: rke-cis-1.7-hardened. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Custom hardened downstream RKE1 cluster with k8s v1.27 and CIS profile v1.23 are created successfully.  </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the cluster successfully. </li>
                    <li> kube-version annotation in Chart's yaml is as expected. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>
        <tr>
            <td>2</td>
            <td>Fresh install Rancher with hardened downstream RKE2 cluster - k8s v1.27 and CIS profile v1.23</td>
            <td>
                <ol>
                    <li> Create a rancher server v2.8-head. </li> 
                    <li> Create a custom hardened downstream RKE2 cluster with k8s v1.27 and CIS profile v1.23. </li>
                    <li> Install "rancher-cis-benchmark" chart on the cluster.</li>
                    <li> Verify kube-version annotation in Chart's yaml: catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.28.0-0' </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: rke2-cis-1.7-hardened. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Custom hardened downstream RKE2 cluster with k8s v1.27 and CIS profile v1.23 are created successfully.  </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the cluster successfully. </li>
                    <li> kube-version annotation in Chart's yaml is as expected. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>
        <tr>
            <td>3</td>
            <td>Fresh install Rancher with hardened downstream K3s cluster - k8s v1.27 and CIS profile v1.23</td>
            <td>
                <ol>
                    <li> Create a rancher server v2.8-head. </li> 
                    <li> Create a custom hardened downstream K3s cluster with k8s v1.27 and CIS profile v1.23. </li>
                    <li> Install "rancher-cis-benchmark" chart on the cluster.</li>
                    <li> Verify kube-version annotation in Chart's yaml: catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.28.0-0' </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: k3s-cis-1.7-hardened. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Custom hardened downstream K3s cluster with k8s v1.27 and CIS profile v1.23 are created successfully.  </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the cluster successfully. </li>
                    <li> kube-version annotation in Chart's yaml is as expected. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>
        <tr>
            <td>4</td>
            <td>Fresh install Rancher with non-hardened downstream RKE1 cluster - k8s v1.27</td>
            <td>
                <ol>
                    <li> Create a rancher server v2.8-head. </li> 
                    <li> Create a custom non-hardened downstream RKE1 cluster with k8s v1.27. </li>
                    <li> Install "rancher-cis-benchmark" chart on the cluster.</li>
                    <li> Verify kube-version annotation in Chart's yaml: catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.28.0-0' </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: rke-cis-1.7-permissive. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Custom non-hardened downstream RKE1 cluster with k8s v1.27 is created successfully.  </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the cluster successfully. </li>
                    <li> kube-version annotation in Chart's yaml is as expected. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>
        <tr>
            <td>5</td>
            <td>Fresh install Rancher with non-hardened downstream RKE2 cluster - k8s v1.27.</td>
            <td>
                <ol>
                    <li> Create a rancher server v2.8-head. </li> 
                    <li> Create a custom non-hardened downstream RKE2 cluster with k8s v1.27. </li>
                    <li> Install "rancher-cis-benchmark" chart on the cluster.</li>
                    <li> Verify kube-version annotation in Chart's yaml: catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.28.0-0' </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: rke2-cis-1.7-permissive. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Custom non-hardened downstream RKE2 cluster with k8s v1.27 is created successfully.  </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the cluster successfully. </li>
                    <li> kube-version annotation in Chart's yaml is as expected. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>
        <tr>
            <td>6</td>
            <td>Fresh install Rancher with non-hardened downstream K3s cluster - k8s v1.27.</td>
            <td>
                <ol>
                    <li> Create a rancher server v2.8-head. </li> 
                    <li> Create a custom non-hardened downstream K3s cluster with k8s v1.27. </li>
                    <li> Install "rancher-cis-benchmark" chart on the cluster.</li>
                    <li> Verify kube-version annotation in Chart's yaml: catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.28.0-0' </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: k3s-cis-1.7-permissive. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Custom non-hardened downstream K3s cluster with k8s v1.27 is created successfully.  </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the cluster successfully. </li>
                    <li> kube-version annotation in Chart's yaml is as expected. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>
        <tr>
            <td>7</td>
            <td>Fresh install Rancher with hardened local cluster - k8s v1.27</td>
            <td>
                <ol>
                    <li> Create a hardened rancher server v2.8-head (on RKE cluster with k8s v1.27). </li> 
                    <li> Install "rancher-cis-benchmark" chart on the local cluster.</li>
                    <li> Verify kube-version annotation in Chart's yaml: catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.28.0-0' </li>
                    <li> Run CIS scan on the local cluster with the following cis benchmark profile: rke-cis-1.7-hardened. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the local cluster successfully. </li>
                    <li> kube-version annotation in Chart's yaml is as expected. </li>
                    <li> CIS scan on the local cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>
        <tr>
            <td>8</td>
            <td>Fresh install Rancher with non-hardened local cluster - k8s v1.27</td>
            <td>
                <ol>
                    <li> Create a non-hardened rancher server v2.8-head (on RKE cluster with k8s v1.27). </li> 
                    <li> Install "rancher-cis-benchmark" chart on the local cluster.</li>
                    <li> Verify kube-version annotation in Chart's yaml: catalog.cattle.io/kube-version: '>= 1.23.0-0 < 1.28.0-0' </li>
                    <li> Run CIS scan on the local cluster with the following cis benchmark profile: rke-cis-1.7-permissive. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the local cluster successfully. </li>
                    <li> kube-version annotation in Chart's yaml is as expected. </li>
                    <li> CIS scan on the local cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>       
        <tr>
            <td>9</td>
            <td>Fresh install Rancher with hardened downstream cluster and non-default CNI</td>
            <td>
                <ol>
                    <li> Create a rancher server v2.8-head. </li> 
                    <li> Create a custom hardened RKE2 downstream cluster (v1.27 with CIS profile v1.23 and canal as CNI) </li>
                    <li> Install "rancher-cis-benchmark" chart on the downstream cluster. </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: rke2-cis-1.7-hardened. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Custom hardened RKE2 downstream cluster is created successfully. </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the downstream cluster successfully. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>
        <tr>
            <td>10</td>
            <td>Rancher, K8s and CIS chart upgrade with hardened downstream cluster </td>
            <td>
                <ol>
                    <li> Create a rancher server on the latest released version. </li> 
                    <li> Create a custom hardened RKE2 downstream cluster (v1.26 with CIS profile v1.23) </li>
                    <li> Install "rancher-cis-benchmark" chart on the downstream cluster.</li>
                    <li> Upgrade Rancher to v2.8-head. </li>
                    <li> Upgrade CIS chart from 4.2.0 to latest version under test. </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: rke2-cis-1.7-hardened. </li>
                    <li> Upgrade K8s from v1.26 to v1.27. </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: rke2-cis-1.7-hardened. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Custom hardened RKE2 downstream cluster is created successfully. </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the downstream cluster successfully. </li>
                    <li> Rancher is upgraded to v2.8-head successfully. </li>
                    <li> CIS chart is upgraded from 4.2.0 to latest version successfully. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> K8s is upgraded from v1.26 to v1.27 successfully. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>
        <tr>
            <td>11</td>
            <td>Rancher, K8s and CIS chart upgrade with non-hardened downstream cluster </td>
            <td>
                <ol>
                    <li> Create a rancher server on the latest released version. </li> 
                    <li> Create a non-hardened RKE2 downstream cluster (K8s v1.26). </li>
                    <li> Install "rancher-cis-benchmark" chart on the downstream cluster.</li>
                    <li> Upgrade Rancher to v2.8-head. </li>
                    <li> Upgrade CIS chart from 4.2.0 to latest version under test. </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: rke2-cis-1.7-permissive. </li>
                    <li> Upgrade K8s from v1.26 to v1.27. </li>
                    <li> Run CIS scan on the downstream cluster with the following cis benchmark profile: rke2-cis-1.7-permissive. </li>
                    <li> Verify that the scan report does not contain any sensitive information. </li>
                    <li> Generate a JSON report and ensure that the report does not contain any errors. </li>
                    <li> Uninstall "rancher-cis-benchmark" chart. </li>
                </ol>
            </td>
            <td> 
                <ol>
                    <li> Rancher installation completes successfully. </li>
                    <li> Non-hardened RKE2 downstream cluster is created successfully. </li>
                    <li> Chart "rancher-cis-benchmark" is installed on the downstream cluster successfully. </li>
                    <li> Rancher is upgraded to v2.8-head successfully. </li>
                    <li> CIS chart is upgraded from 4.2.0 to latest version successfully. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> K8s is upgraded from v1.26 to v1.27 successfully. </li>
                    <li> CIS scan on the downstream cluster pass. </li>
                    <li> Scan report does not contain any sensitive information. </li>
                    <li> JSON report does not contain any errors. </li>
                    <li> Chart "rancher-cis-benchmark" is uninstalled successfully.</li>
                </ol>
            </td>
        </tr>
    </tbody>
</table>

**Comment by rayandas** [comment]
Dev testing is done with K8s 1.27.5 on Rancher v2.8-head. 
![Screenshot from 2023-09-14 15-11-01](https://github.com/rancher/rancher/assets/32811240/e38f0894-1b95-40d4-97fa-e0c405da6e8b)
![Screenshot from 2023-09-14 15-12-31](https://github.com/rancher/rancher/assets/32811240/657f0330-3cfe-41fb-ae0b-0526dea1cd1f)

**Comment by Priyashetty17** [comment]
@rayandas Moving this issue to Blocked since we are waiting on https://github.com/rancher/charts/pull/2988#event-10380480219 to get merged cc @andypitcher

**Comment by rayandas** [comment]
@Priyashetty17 https://github.com/rancher/charts/pull/2988 is now merged.

**Comment by Priyashetty17** [comment]
Fresh install checks passed with [Chart 5.0.0-rc2](https://github.com/rancher/charts/tree/dev-v2.8/charts/rancher-cis-benchmark/5.0.0-rc2). Rancher upgrade scenario is pending and is currently blocked with: 
https://github.com/rancher/rancher/issues/42888

**Comment by gunamata** [comment]
@Priyashetty17 , Given https://github.com/rancher/rancher/issues/42888 is closed, Are you able to test the Rancher upgrade scenario?

**Comment by Priyashetty17** [comment]
Yes @gunamata , I plan to complete this today.

**Comment by Priyashetty17** [comment]
Validated with v2.8-741111372fad99c30070359990314c6dcb6707b3-head, CIS chart: 5.0.0-rc3
